### PR TITLE
New Genesis Template cli command

### LINKED ISSUE #580: Add CLI Command to Generate New GenesisTemplate
https://github.com/filecoin-project/lotus/blob/9b1d2f4c137433c1fe1f5ed44ba2e70105a409f0/cmd/lotus-seed/genesis.go#L29

### PR DESCRIPTION
**Pull request by snaumov**
**Summary of changes**
Changes introduced in this pull request:
- 

Added new CLI command to create Genesis template. 

Lotus-seed output for reference:

```
{
  "Accounts": [],
  "Miners": [],
  "NetworkName": "localnet-a793c73a-4db3-4755-910e-7466bce48023"
}
```

Forest's output:

```
{
  "Accounts": [],
  "Miners": [],
  "NetworkName": "localnet-a793c73a-4db3-4755-910e-7466bce4blah"
}
```

**How to test**

build and run

`forest genesis new-template`

Provide network name and output path optionally:

`forest genesis new-template -n "my-network" -f "./genesis_out.json"`

**Reference issue to close (if applicable)**

Closes #580 #579 

**Other information and links**

### CODE REVIEW

**Comment by ec2** [comment]
One general note so far, is i'd add this attribute to the serde-able structs to get pascal casing so we can match Lotus `#[serde(rename_all = "PascalCase")]`

**Comment by snaumov** [comment]
> One general note so far, is i'd add this attribute to the serde-able structs to get pascal casing so we can match Lotus `#[serde(rename_all = "PascalCase")]`. Also thanks so much for the contribution!!! It's always so exciting getting external contributors

Added `PascalCase`.

Ya, absolutely, the pleasure is all mine :+1:

File: types/src/genesis/mod.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::SectorSize;
+use address::Address;
+use num_bigint::bigint_ser;
+use serde::Serialize;
+use vm::TokenAmount;
+
+#[derive(Serialize)]
```
Comment (austinabell) [inline types/src/genesis/mod.rs:10]: Optionally just add a TODO that this won't serialize and deserialize as a string (default enum serialization is as a map)

File: types/src/genesis/mod.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+pub struct Actor {
+    actor_type: ActorType,
+    #[serde(with = "bigint_ser")]
+    token_amount: TokenAmount,
+}
+
+#[derive(Serialize)]
+pub struct Miner {
```
Comment (austinabell) [inline types/src/genesis/mod.rs:24]: can you add pascal casing to these as well?

File: types/src/genesis/mod.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    sector_size: SectorSize,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "PascalCase")]
+pub struct Template {
+    accounts: Vec<Actor>,
+    miners: Vec<Miner>,
+    network_name: String,
+    // timestamp: SystemTime,
```
Comment (austinabell) [inline types/src/genesis/mod.rs:42]: these fields should probably be public (no reason not to for something like this). This also applies to the structures above as well

File: types/src/genesis/mod.rs (hunk 2 above)
Comment (snaumov) [reply types/src/genesis/mod.rs:24]: added

File: types/src/genesis/mod.rs (hunk 3 above)
Comment (snaumov) [reply types/src/genesis/mod.rs:42]: made public

File: types/src/genesis/mod.rs (hunk 1 above)
Comment (snaumov) [reply types/src/genesis/mod.rs:10]: hmm, but my intention was for this enum to serialize/deserialize as a string.
I'm happy to resolve this TODO now, are you saying I need to changing something?

Comment (austinabell) [reply types/src/genesis/mod.rs:10]: Yeah, you can do so by manually implementing Serialize (and Deserialize if you intend to) and just serialize it how you want. I don't think there is anything you can tag an enum to achieve this specifically. 

for ref: https://serde.rs/enum-representations.html

Comment (snaumov) [reply types/src/genesis/mod.rs:10]: hmm, testing now and it seems to serialize fine as a string: 

```
{
  "Accounts": [],
  "Miners": [],
  "NetworkName": "localnet-bc9243f8-5ea9-4bd0-a96c-9eb471ac3872",
  "ActorType": "Account" // <- THIS
}
```

(I temporarily changed `Template` struct for testing)

```
#[derive(Serialize)]
#[serde(rename_all = "PascalCase")]
pub struct Template {
    pub accounts: Vec<Actor>,
    pub miners: Vec<Miner>,
    pub network_name: String,
    actor_type: ActorType,
    // timestamp: SystemTime,
}

impl Template {
    pub fn new(network_name: String) -> Template {
        Template {
            accounts: Vec::new(),
            miners: Vec::new(),
            network_name,
            actor_type: ActorType::Account,
        }
    }
}
```

Comment (austinabell) [reply types/src/genesis/mod.rs:10]: oh, my bad then, I didn't expect that to happen for cases with no data. Perhaps with this case you should make it all lower case to match their format as this is their strings:
```
	TAccount  ActorType = "account"
	TMultisig ActorType = "multisig"
```

**Comment by snaumov** [comment]
@austinabell @ec2 please merge when you have a chance. Thanks :D

**Comment by austinabell** [comment]
@snaumov thank you !

**Pull request merged**